boot: Add BCD store parser

medhefgo opened the pull request:
I guess the Windows 11 upgrade gave me the motivation to write a proper BCD parser. Turns out writing one isn't that hard to do. :joy_cat:

This was prototyped in userspace with all kinds of BCD stores acquired from VMs and so I am quite confident that it does the job right. If desired I can easily add a userspace unit test for this. Though, it would require adding at least one BCD store to the source tree and I am not sure how we feel about that. :thinking:

bluca commented:
> Though, it would require adding at least one BCD store to the source tree and I am not sure how we feel about that.

How big would that be, compressed? We have fuzzing binaries already in the repo, as long as it never changes and it's not too big, it's ok

medhefgo commented:
A typical store is about 20-30KiB and goes down to ~4KiB xz compressd.

bluca commented:
> A typical store is about 20-30KiB and goes down to ~4KiB xz compressd.

4KB compressed would be perfectly fine - again as long as it never changes

keszybz commented:
The code looks good. But please include the test cases. We really should avoid adding new functionality without any test code: it just makes things so much easier if the code is exercised at least a bit. But please use zstd for compression: the difference in size is not important, but running the tests will be noticeably faster.

keszybz commented:
@medhefgo ping. Please submit an updated version, I'd love to see this in v250 and we're getting very close to a pre-release.

medhefgo commented:
> @medhefgo ping. Please submit an updated version, I'd love to see this in v250 and we're getting very close to a pre-release.

It should be done by tomorrow. Writing a unit test actually revealed some subtle bugs in offset/size sanity checks.

medhefgo commented:
The code should now hopefully catch anything you throw at it. I even found a sysinternals tool to compact the reg hives (while also removing any keys/values we don't care for), so they are extra small.

bluca commented:
Can you mark the binary files as `binary` in `.gitattributes` please?

keszybz commented:
ci/semaphoreci: debian autopkgtest (lxc) is still unhappy:
```
autopkgtest [15:59:54]: test unit-tests: [-----------------------
[31mFAIL: test-bcd[0m
/* test_argv_bcds */
/* test_bad_bcd */
Assertion 'read_full_file_full(AT_FDCWD, fn, UINT64_MAX, SIZE_MAX, 0, NULL, &compressed, &len) >= 0' failed at src/boot/efi/test-bcd.c:19, function load_bcd(). Aborting.
```

medhefgo commented:
> ci/semaphoreci: debian autopkgtest (lxc) is still unhappy:

I think that's because I haven't put an install rule for the bcd test data. Not sure if I should do that or just skip the test if the data cannot be found...

keszybz commented on src/boot/efi/bcd.c:
```diff
@@ -0,0 +1,286 @@
+        UINT32 sig;
+        UINT32 primary_seqnum;
+        UINT32 secondary_seqnum;
+        UINT64 _pad1;
```
Hmm, uint64_t would be aligned to 64 bits. Here `_packed_` is used, so the alignment does not happen. It still feels strange to have a purposefully unaligned field.

medhefgo replied:
All these structs have way more members than we use. I just chose to squash them all into _padN members instead of spelling them all out. I can drop a comment for this. (Or I am just not quite getting your meaning here :D)

(This one would actually normally be a uint64_t timestamp though.)

keszybz replied:
OK, ignore my comment then.

keszybz commented on src/boot/efi/bcd.c:
```diff
@@ -0,0 +1,286 @@
+        if (key->sig != SIG_KEY)
+                return NULL;
+
+        if ((UINT64) offset + offsetof(Key, key_name) + sizeof(CHAR8[key->key_name_len]) > bcd_len)
```
> `sizeof(CHAR8[key->key_name_len])`

~~That's just `key->key_name_len`. Why not use that directly?~~

OK, I see this is done for consistency with `sizeof(INT32[...])` below. Makes sense.

keszybz commented on src/boot/efi/bcd.c:
```diff
@@ -0,0 +1,286 @@
+        assert(bcd);
+        assert(name);
+
+        if ((UINT64) offset + sizeof(Key) > bcd_len)
```
This cast is not necessary — casting to the wider type happens automatically.

medhefgo replied:
Afaik, sizeof(size_t) is 32bit on 32bit machines, no? That's why I put in the casts.

keszybz replied:
Indeed, I didn't think about that.

keszybz commented on src/boot/efi/bcd.c:
```diff
@@ -0,0 +1,286 @@
+
+static Key *get_key(const UINT8 *bcd, UINT32 bcd_len, UINT32 offset, const CHAR8 *name);
+
+static Key *get_subkey(const UINT8 *bcd, UINT32 bcd_len, UINT32 offset, const CHAR8 *name) {
```
Is there a limit to nesting level here? If the `name` is under our control, then that should be fine. We just shouldn't use this to look up keys that are specified by the parsed data, because we might recurse too deeply.

Maybe add a comment mentioning this, to avoid a mistake later.

keszybz commented on src/boot/efi/bcd.c:
```diff
@@ -0,0 +1,286 @@
+assert_cc(offsetof(KeyValue, data_type) == 12);
+assert_cc(offsetof(KeyValue, name) == 20);
+
+static Key *get_key(const UINT8 *bcd, UINT32 bcd_len, UINT32 offset, const CHAR8 *name);
```
I know this style is widely used, but please consider moving the star adjacent to the return base type:
```C
static Key* get_key(…);
```
I think that's more readable, because we're returning a `Key*` and now it's all in one place.

--

Since the input data is `const`, this should be `const Key*` too, here and below.

keszybz commented on src/boot/efi/util.h:
```diff
@@ -130,6 +130,7 @@ EFI_STATUS readdir_harder(EFI_FILE_HANDLE handle, EFI_FILE_INFO **buffer, UINTN
 
 UINTN strnlena(const CHAR8 *p, UINTN maxlen);
 CHAR8 *xstrndup8(const CHAR8 *p, UINTN sz);
+INTN strncasecmpa(const CHAR8 *a, const CHAR8 *b, UINTN maxlen);
```
We generally don't use strcmp() and friends in the codebase, but wrap it in `streq()`. Please also add `static inline strncaseeqa()` here.

keszybz commented on src/boot/efi/bcd.c:
```diff
@@ -0,0 +1,286 @@
+                return NULL;
+
+        if (*name) {
+                if (strncasecmpa(name, key->key_name, key->key_name_len) == 0 && !name[key->key_name_len])
```
`strncaseeqa()` here and in other places…

keszybz commented on src/boot/efi/bcd.c:
```diff
@@ -0,0 +1,286 @@
+                if (kv->sig != SIG_KEY_VALUE)
+                        continue;
+
+                if ((UINT64) offset + offsetof(KeyValue, name) + kv->name_len> bcd_len)
```
Missing space before ">".

keszybz commented on src/boot/efi/bcd.c:
```diff
@@ -0,0 +1,286 @@
+                /* If most significant bit is set, data is stored in data_offset itself, but
+                 * we are only interested in UTF16 strings. The only strings that could fit
+                 * would have just one char in it, so let's not bother with this. */
+                if (FLAGS_SET(kv->data_size, UINT32_C(1) << UINT32_C(31)))
```
That second UINT32_C() is unnecessary and specifying the width is pretty much pointless: it has no effect on the output.

keszybz commented on src/boot/efi/bcd.c:
```diff
@@ -1,8 +1,28 @@
-#include "macro-fundamental.h"
-#include "util.h"
+#ifdef SD_BOOT
+#        include <efi.h>
```
We use two spaces for indentation of `#  include` and `#  define`.

medhefgo replied:
Does clang-format really not have an option for this? *sigh*

keszybz commented on src/boot/efi/bcd.c:
```diff
@@ -202,7 +222,7 @@ static KeyValue *get_key_value(const UINT8 *bcd, UINT32 bcd_len, const Key *key,
  * one GUID, the BCD is multi-boot and we stop looking. Otherwise we take that GUID, look it
  * up, and return its description property. */
-CHAR16 *get_bcd_title(UINT8 *bcd, UINTN bcd_len) {
+HOSTESD_STATIC CHAR16 *get_bcd_title(UINT8 *bcd, UINTN bcd_len) {
```
That seems to be a typo here.

keszybz reviewed (commented):
This looks very nice.

medhefgo commented:
Except for the "Key*" thing, I addressed all issues. I really hate that pointer style and am not changing that unless you insist.

I also added verification of the found GUID along with a test case for it. Dunno why I did not think of that myself.

yuwata commented on src/boot/efi/boot.c:
```diff
@@ -1954,34 +1955,8 @@ static void config_entry_add_windows(Config *config, EFI_HANDLE *device, EFI_FIL
-                }
-        }
+        if (!EFI_ERROR(err))
+                title = get_bcd_title((UINT8 *) bcd, len);
```
Why this cast is necessary?

medhefgo replied:
Fedora again:
`error: pointer targets in passing argument 1 of "get_bcd_title" differ in signedness`

yuwata replied:
Ah, `UINT8*` vs `CHAR8*`. Thanks.

yuwata reviewed (approved):
Not familiar with this, but superficially LGTM.

yuwata commented on src/boot/efi/bcd.c:
```diff
@@ -0,0 +1,321 @@
+#endif
+
+enum {
+        SIG_BASE_BLOCK = 1718052210, /* regf */
+        SIG_KEY = 27502,             /* nk */
+        SIG_SUBKEY_FAST = 26220,     /* lf */
+        SIG_KEY_VALUE = 27510,       /* vk */
```
I prefer that these are aligned.

yuwata commented on src/boot/efi/bcd.c:
```diff
@@ -0,0 +1,321 @@
+};
+
+enum {
+        REG_SZ = 1,
+        REG_MULTI_SZ = 7,
```
Also here.

medhefgo commented:
Something I forgot to ask so far: Is there a particular reason we compile with `-Wpointer-arith`? I really like gcc's extension that allows treating `void *` like `char *` when it comes to pointer arithmetic, making so many casts unnecessary.
